File: modules/plug/Plug.py
############################################################################
##
## Copyright (c) 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009,
## 2010, 2011 BalaBit IT Ltd, Budapest, Hungary
##
## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 2 of the License, or
## (at your option) any later version.
##
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with this program; if not, write to the Free Software
## Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
##
##
## $Id: Plug.py,v 1.32 2004/06/15 12:28:47 sasa Exp $
##
## Author  : Bazsi
## Auditor : kisza
## Last audited version: 1.10
## Notes:
##
############################################################################

"""<module maturity="stable">
  <summary>
    Proxy for transferring data without protocol inspection.
  </summary>
  <description>
    <para>
      This module defines an interface to the Plug proxy. Plug is a simple TCP or
      UDP circuit, which means that transmission takes place without protocol
      verification.
    </para>
    <section>
      <title>Proxy behavior</title>
      <para>
        This class implements a general plug proxy, and is capable of optionally
        disabling data transfer in either direction. Plug proxy reads
        connection on the client side, then creates another connection at the 
        server side. Arriving responses are sent back to the client. However, it is not a protocol proxy, therefore PlugProxy does not implement any protocol analysis. It offers protection to clients and servers from lower level (e.g.: IP) attacks. It is mainly used to allow traffic pass the firewall for which there is no protocol proxy available.
      </para>
      <para>
        By default plug copies all data in both directions. To change
        this behavior, set the <parameter>copy_to_client</parameter> or <parameter>copy_to_server</parameter> attribute
        to FALSE.
      </para>
    <para>
      Plug supports the use of secondary sessions as described in <xref linkend="secondary_sessions">secondary sessions</xref>.
      </para>
      <note>
          <para>
            Copying of out-of-band data is not supported.
          </para>
      </note>
      </section>
    <section>
      <title>Related standards</title>
        <para>
          Plug proxy is not a protocol specific proxy module, therefore
          it is not specified in standards.
        </para>
    </section>
  </description>
  <metainfo>
    <constants>
      <constantgroup maturity="stable" id="const.plug.log">
        <description>
          Log level defined in Plug module 
        </description>
        <item>
          <name>PLUG_DEBUG</name>
          <value>"plug.debug"</value>
        </item>
      </constantgroup>
    </constants>
  </metainfo>
</module>
"""

from Zorp import *
from Proxy import Proxy

PLUG_DEBUG = "plug.debug"

class AbstractPlugProxy(Proxy):
	"""<class maturity="stable" abstract="yes">
          <summary>
            Class encapsulating the abstract Plug proxy.
          </summary>
          <description>
	  <para>An abstract proxy class for transferring data.</para>
          </description>
          <metainfo>
            <attributes>
              <attribute maturity="stable">
                <name>copy_to_server</name>
                <type>
                  <boolean/>
                </type>
                <default>TRUE</default>
                <conftime>
                  <write/>
                </conftime>
                <runtime>
                  <read/>
                </runtime>
                <description>
                  Allow data transfer in the client->server direction.
                </description>
              </attribute>
              <attribute maturity="stable">
                <name>copy_to_client</name>
                <type>
                  <boolean/>
                </type>
                <default>TRUE</default>
                <conftime>
                  <write/>
                </conftime>
                <runtime>
                  <read/>
                </runtime>
                <description>
                  Allow data transfer in the server->client direction.
                </description>
              </attribute>
              <attribute maturity="stable">
                <name>bandwidth_to_client</name>
                <type>
                  <integer/>
                </type>
                <default>n/a</default>
                <conftime/>
                <runtime>
                  <read/>
                </runtime>
                <description>
                  Read-only variable containing the bandwidth currently used in server->client direction.
                </description>
              </attribute>
              <attribute maturity="stable">
                <name>bandwidth_to_server</name>
                <type>
                  <integer/>
                </type>
                <default>n/a</default>
                <conftime/>
                <runtime>
                  <read/>
                </runtime>
                <description>
                  Read-only variable containing the bandwidth currently used in client->server direction.
                </description>
              </attribute>
              <attribute maturity="stable">
                <name>packet_stats_interval_time</name>
                <type>
                  <integer/>
                </type>
                <default>0</default>
                <conftime>
                  <write/>
                </conftime>
                <runtime>
                  <read/>
                </runtime>
                <description>
                  The time in milliseconds between two successive packetStats() events.
		  It can be useful when the Quality of Service for the connection is influenced dynamically.
                  Set to 0 to turn packetStats() off.
                </description>
              </attribute>
              <attribute maturity="stable">
                <name>packet_stats_interval_packet</name>
                <type>
                  <integer/>
                </type>
                <default>0</default>
                <conftime>
                  <write/>
                </conftime>
                <runtime>
                  <read/>
                </runtime>
                <description>
                  The number of passing packages between two successive packetStats() events.  
		  It can be useful when the Quality of Service for the connection is influenced dynamically.
                  Set to 0 to turn packetStats() off.
                </description>
              </attribute>
              <attribute maturity="stable">
                <name>stack_proxy</name>
		<type>
		  <link id="action.zorp.stack"/>
		</type>
                <conftime>
                  <write/>
                </conftime>
                <runtime>
                  <read/>
                </runtime>
                <description>
                   Proxy class to stack into the connection. All data is passed to the specified proxy.
                </description>
              </attribute>
              <attribute maturity="stable">
                <name>timeout</name>
                <type>
                  <integer/>
                </type>
                <default>60000</default>
                <conftime>
                  <write/>
                </conftime>
                <runtime>
                  <read/>
                </runtime>
                <description>
                  I/O timeout in milliseconds.
                </description>
              </attribute>
              <attribute maturity="stable">
                <name>shutdown_soft</name>
                <type>
                  <boolean/>
                </type>
                <default>FALSE</default>
                <conftime>
                  <write/>
                </conftime>
                <runtime>
                  <read/>
                </runtime>
                <description>
                  If enabled, the two sides of a connection are closed separately. (E.g.: if the server closes the connection the client side connection is held until it is verified that no further data arrives, for example from a stacked proxy.) It is automatically enabled when proxies are stacked into the connection.
                </description>
              </attribute>
              <attribute maturity="stable">
                <name>buffer_size</name>
                <type>
                  <integer/>
                </type>
                <default>1500</default>
                <conftime>
                  <write/>
                </conftime>
                <runtime>
                  <read/>
                </runtime>
                <description>
                  Size of the buffer used for copying data.
                </description>
              </attribute>
              <attribute maturity="stable">
                <name>secondary_sessions</name>
                <type>
                  <integer/>
                </type>
                <default>10</default>
                <conftime>
                  <read/>
                  <write/>
                </conftime>
                <runtime>
                  <read/>
                </runtime>
                <description>
                  Maximum number of allowed secondary sessions within a single proxy instance. See <xref linkend="secondary_sessions"/> for details.
                </description>
              </attribute>
              <attribute maturity="stable">
                <name>secondary_mask</name>
                <type>
                  <secondary_mask/>
                </type>
                <default>0xf</default>
                <conftime>
                  <read/>
                  <write/>
                </conftime>
                <runtime>
                  <read/>
                </runtime>
                <description>
                  Specifies which connections can be handled by the same proxy instance. See <xref linkend="secondary_sessions"/> for details.
                </description>
              </attribute>
            </attributes>
          </metainfo>
        </class>
        """
	name = "plug"
	def __init__(self, session):
		"""<method internal="yes">
                  <summary>
                    Constructor initializing a PlugProxy instance.
                  </summary>
                  <description>
                    <para>
                      This constructor creates and sets up a PlugProxy instance.
                    </para>
                  </description>
                  <metainfo>
                    <arguments>
                      <argument maturity="stable">
                        <name>session</name>
                        <type>SESSION</type>
                        <description>
                          session this instance belongs to
                        </description>
                      </argument>
                    </arguments>
                  </metainfo>
                </method>
                """
		self.stack_proxy = None
		Proxy.__init__(self, session)

	def requestStack(self):
		"""<method internal="yes">
                  <summary>
                    Function returning the stacked proxy class.
                  </summary>
                  <description>
                    <para>
		      This callback is called by the underlying C proxy to query
		      if something is to be stacked into it. It should return
		      the proxy class to be used.
		      Returns the class of the proxy to stack.
                    </para>
                  </description>
                  <metainfo>
                    <arguments/>
                  </metainfo>
                </method>
                """
		return self.stack_proxy

	def packetStats(self, client_bytes, client_pkts, server_bytes, server_pkts):
		"""<method maturity="stable">
                  <summary>
                    Function called when the packet_stats_interval is elapsed.
                  </summary>
                  <description>
                    <para>
                     This function is called whenever the time interval set in packet_stats_interval elapses, or a given number of
                     packets were transmitted. This event receives packet
                     statistics as parameters.
                    </para>
                    <para>
                     This function can be used in managing the Quality of Service of the connections; e.g.: to terminate connections with excessive
                     bandwidth requirements (for instance to limit the impact of
                     a covert channel opened when using plug instead of a                    protocol specific proxy).
                     </para>
                  </description>
                  <metainfo>
                    <arguments>
                      <argument maturity="stable">
                        <name>client_bytes</name>
                        <type></type>
                        <description>
                          Number of bytes transmitted to the client.
                        </description>
                      </argument>
                      <argument maturity="stable">
                        <name>client_pkts</name>
                        <type></type>
                        <description>
                          Number of packets transmitted to the client.
                        </description>
                      </argument>
                      <argument maturity="stable">
                        <name>server_bytes</name>
                        <type></type>
                        <description>
                          Number of bytes transmitted to the server.
                        </description>
                      </argument>
                      <argument maturity="stable">
                        <name>server_pkts</name>
                        <type></type>
                        <description>
                          Number of packets transmitted to the server.
                       </description>
                      </argument>
                    </arguments>
                  </metainfo>
                </method>
                """
		raise NotImplementedError


class PlugProxy(AbstractPlugProxy):
	"""<class maturity="stable">
          <summary>
            Class encapsulating the default Plug proxy.
          </summary>
          <description>
            <para>
              A default PlugProxy based on AbstractPlugProxy.
            </para>
          </description>
          <metainfo>
            <attributes/>
          </metainfo>
        </class>
        """
	pass

File: modules/plug/plug.c
/***************************************************************************
 *
 * Copyright (c) 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009,
 * 2010, 2011 BalaBit IT Ltd, Budapest, Hungary
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 as published
 * by the Free Software Foundation.
 *
 * Note that this permission is granted for only version 2 of the GPL.
 *
 * As an additional exemption you are allowed to compile & link against the
 * OpenSSL libraries as published by the OpenSSL project. See the file
 * COPYING for details.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * $Id: plug.c,v 1.89 2004/07/22 09:47:36 bazsi Exp $
 *
 * Author: Balazs Scheidler <bazsi@balabit.hu>
 * Auditor:
 * Last audited version:
 * Notes:
 *
 ***************************************************************************/

#include <zorp/plugsession.h>
#include <zorp/thread.h>
#include <zorp/streamfd.h>
#include <zorp/proxy.h>
#include <zorp/poll.h>
#include <zorp/policy.h>
#include <zorp/thread.h>
#include <zorp/log.h>
#include <zorp/registry.h>
#include <zorp/sockaddr.h>
#include <zorp/io.h>

#include <sys/types.h>
#include <sys/time.h>
#include <unistd.h>
#include <errno.h>
#include <arpa/inet.h>

#define PLUG_DEFAULT_BUFSIZE    1500

#define PLUG_DEBUG "plug.debug"
#define PLUG_ERROR "plug.error"
#define PLUG_POLICY "plug.policy"
#define PLUG_SESSION "plug.session"

typedef struct _PlugProxy
{
  ZProxy super;
  ZPoll *poll;
  ZPlugSessionData session_data;
  ZPlugSession *session;
} PlugProxy;

extern ZClass PlugProxy__class;
            
static gboolean
plug_packet_stat_event(ZPlugSession *session,
                       guint64 client_bytes, guint64 client_pkts, 
                       guint64 server_bytes, guint64 server_pkts,
                       gpointer user_data);
static void
plug_finish(ZPlugSession *session, gpointer user_data);
                       
void
plug_config_set_defaults(PlugProxy *self)
{
  z_proxy_enter(self);

  self->session_data.copy_to_server = TRUE;
  self->session_data.copy_to_client = TRUE;
  self->session_data.timeout = 600000;
  self->session_data.buffer_size = PLUG_DEFAULT_BUFSIZE;
  self->session_data.packet_stats = plug_packet_stat_event;
  self->session_data.finish = plug_finish;
  
  if (self->super.parent_proxy)
    self->session_data.shutdown_soft = TRUE;

  z_proxy_leave(self);
}

void
plug_register_vars(PlugProxy *self)
{
  z_proxy_enter(self);

  z_proxy_var_new(&self->super,
                  "timeout",
                  Z_VAR_GET | Z_VAR_SET_CONFIG | Z_VAR_TYPE_INT,
                  &self->session_data.timeout);

  z_proxy_var_new(&self->super,
                  "copy_to_client",
                  Z_VAR_GET | Z_VAR_SET_CONFIG | Z_VAR_TYPE_INT,
                  &self->session_data.copy_to_client);

  z_proxy_var_new(&self->super,
                  "copy_to_server",
                  Z_VAR_GET | Z_VAR_SET_CONFIG | Z_VAR_TYPE_INT,
                  &self->session_data.copy_to_server);

  z_proxy_var_new(&self->super,
                  "shutdown_soft",
                  Z_VAR_GET | Z_VAR_SET_CONFIG | Z_VAR_TYPE_INT,
                  &self->session_data.shutdown_soft);

  z_proxy_var_new(&self->super,
                  "packet_stats_interval_packet",
                  Z_VAR_GET | Z_VAR_SET_CONFIG | Z_VAR_TYPE_INT,
                  &self->session_data.packet_stats_interval_packet);

  z_proxy_var_new(&self->super,
                  "packet_stats_interval_time",
                  Z_VAR_GET | Z_VAR_SET_CONFIG | Z_VAR_TYPE_INT,
                  &self->session_data.packet_stats_interval_time);

  z_proxy_var_new(&self->super,
                  "buffer_size",
                  Z_VAR_GET | Z_VAR_SET_CONFIG | Z_VAR_TYPE_INT,
                  &self->session_data.buffer_size);

  /* Zorp 1.4 compatibility */
  z_proxy_var_new(&self->super,
                  "packet_stats_interval",
                  Z_VAR_TYPE_ALIAS | Z_VAR_GET | Z_VAR_SET | Z_VAR_GET_CONFIG | Z_VAR_SET_CONFIG,
                  "packet_stats_interval_packet");
  z_proxy_return(self);
}

static gboolean
plug_packet_stat_event(ZPlugSession *session G_GNUC_UNUSED,
                       guint64 client_bytes, guint64 client_pkts, 
                       guint64 server_bytes, guint64 server_pkts,
                       gpointer user_data)
{
  PlugProxy *self = (PlugProxy *) user_data;
  ZPolicyObj *res;
  gboolean called;
  guint resc;

  z_policy_lock(self->super.thread);
  res = z_policy_call(self->super.handler, "packetStats",
                      z_policy_var_build("iiii",
                                         (guint32) client_bytes,
                                         (guint32) client_pkts,
                                         (guint32) server_bytes,
                                         (guint32) server_pkts),
                      &called,
                      self->super.session_id);

  if (called)
    {
      resc = Z_REJECT;
      if (res)
        {
          if (!z_policy_var_parse(res, "i", &resc))
            {
              /*LOG
                This message is logged when the policy layer returned a
                non-integer value in its packetStats() function. packetStats()
                is expected to return Z_REJECT or Z_ACCEPT.
               */
              z_proxy_log(self, PLUG_POLICY, 1, "Invalid return value of packetStats(), integer required;");
            }
          else if (resc != Z_ACCEPT)
            {
              /*LOG
                This message indicates that the verdict returned by the
                packetStats() function requests to terminate the session.
               */
              z_proxy_log(self, PLUG_POLICY, 1, "packetStats() requested to abort session; verdict='%d'", resc);
            }
        }
    }
  else
    {
      resc = Z_ACCEPT;
    }
  z_policy_var_unref(res);
  z_policy_unlock(self->super.thread);
  return resc == Z_ACCEPT;
}

static void
plug_finish(ZPlugSession *session G_GNUC_UNUSED, gpointer user_data)
{
  PlugProxy *self = (PlugProxy *) user_data;
  
  z_proxy_nonblocking_stop(&self->super);
}

static gboolean
plug_request_stack_event(PlugProxy *self, ZStackedProxy **stacked)
{
  ZPolicyObj *res;
  gboolean called;
  gboolean rc = TRUE;

  z_proxy_enter(self);
  z_policy_lock(self->super.thread);
  *stacked = NULL;
  res = z_policy_call(self->super.handler,
                      "requestStack",
                      NULL,
                      &called,
                      self->super.session_id);
  if (res)
    {
      if (res != z_policy_none)
        rc = z_proxy_stack_object(&self->super, res, stacked, NULL);
    }
  else if (called)
    {
      rc = FALSE;
    }
  z_policy_var_unref(res);
  z_policy_unlock(self->super.thread);
  z_proxy_return(self, rc);
}

static gboolean
plug_nonblocking_init(ZProxy *s, ZPoll *poll)
{
  PlugProxy *self = Z_CAST(s, PlugProxy);
  ZStackedProxy *stacked;
  
  z_proxy_enter(self);
  
  if (!z_proxy_connect_server(&self->super, NULL, 0))
    {
      z_proxy_leave(self);
      return FALSE;
    }
    
  if (!plug_request_stack_event(self, &stacked))
    {
      z_proxy_leave(self);
      return FALSE;
    }
  
  self->session = z_plug_session_new(&self->session_data, self->super.endpoints[EP_CLIENT], self->super.endpoints[EP_SERVER], stacked, &self->super);
  if (!self->session)
    {
      z_proxy_leave(self);
      return FALSE;
    }
  z_plug_session_register_vars(self->session, self->super.dict);
  
  if (!z_plug_session_start(self->session, poll))
    {
      z_proxy_leave(self);
      return FALSE;
    }
    
  z_proxy_leave(self);
  return TRUE;
}

static void
plug_nonblocking_deinit(ZProxy *s)
{
  PlugProxy *self = Z_CAST(s, PlugProxy);
  
  if (self->session)
    z_plug_session_cancel(self->session);
}

static gboolean
plug_config(ZProxy *s)
{
  PlugProxy *self = (PlugProxy *) s;

  z_proxy_enter(self);

  plug_config_set_defaults(self);
  plug_register_vars(self);
  if (Z_SUPER(s, ZProxy)->config(s))
    z_proxy_return(self, TRUE);

  z_proxy_return(self, FALSE);
}

ZProxy *
plug_proxy_new(ZProxyParams *params)
{
  PlugProxy *self;
  
  z_enter();
  self = Z_CAST(z_proxy_new(Z_CLASS(PlugProxy), params), PlugProxy);
  self->super.flags |= ZPF_NONBLOCKING;
  
  z_leave();
  return &self->super;
}
    
static void
plug_proxy_free(ZObject *s)
{
  PlugProxy *self = Z_CAST(s, PlugProxy);
  
  z_proxy_enter(self);
  
  z_plug_session_destroy(self->session);
  if (self->poll)
    {
      z_poll_unref(self->poll);
      self->poll = NULL;
    }
    
  z_proxy_free_method(s);
  z_return();
}

ZProxyFuncs plug_proxy_funcs =
{
  { 
    Z_FUNCS_COUNT(ZProxy),
    plug_proxy_free,
  },
  .config = plug_config,
  .startup = NULL,
  .nonblocking_init = plug_nonblocking_init,
  .nonblocking_deinit = plug_nonblocking_deinit,
  .shutdown = NULL,
  .destroy = NULL,
};

ZClass PlugProxy__class = 
{
  Z_CLASS_HEADER,
  &ZProxy__class,
  "PlugProxy",
  sizeof(PlugProxy),
  &plug_proxy_funcs.super
};

gint
zorp_module_init(void)
{
  z_registry_add("plug", ZR_PROXY, plug_proxy_new);
  return TRUE;
}
